File: backend/src/application/mappers/product.mapper.ts
import { Product } from "../../domain/entities/product.entity";
import { ProductResponseDto } from "../dto/product-response.dto";

export function toProductResponse(product: Product): ProductResponseDto {
  return {
    id: product.id,
    sku: product.sku,
    name: product.name,
    description: product.description ?? null,
    priceCents: product.priceCents,
    currency: product.currency,
    status: product.status,
    metadata: product.metadata ? { ...product.metadata } : null,
    categories: product.categories.map((category) => ({
      id: category.id,
      name: category.name,
    })),
    type: product.type,
    createdAt: product.createdAt.toISOString(),
    updatedAt: product.updatedAt.toISOString(),
  };
}

File: go-microservice/internal/application/mappers/product_mapper.go
package mappers

import (
	"github.com/southern-martin/go-microservice/internal/application/dto"
	"github.com/southern-martin/go-microservice/internal/domain/entities"
)

func ProductToResponse(product *entities.Product) *dto.ProductResponse {
	if product == nil {
		return nil
	}

	response := &dto.ProductResponse{
		ID:          product.ID,
		Name:        product.Name,
		Description: product.Description,
		SKU:         product.SKU,
		Price:       product.Price,
		Currency:    product.Currency,
		Status:      product.Status,
		CategoryID:  product.CategoryID,
		CreatedAt:   product.CreatedAt,
		UpdatedAt:   product.UpdatedAt,
	}

	// Map category if present
	if product.Category != nil {
		response.Category = &dto.CategoryInfo{
			ID:   product.Category.ID,
			Name: product.Category.Name,
		}
	}

	return response
}

func ProductsToResponse(products []*entities.Product) []dto.ProductResponse {
	responses := make([]dto.ProductResponse, len(products))
	for i, product := range products {
		responses[i] = *ProductToResponse(product)
	}
	return responses
}

func CreateProductRequestToEntity(req *dto.CreateProductRequest) *entities.Product {
	return &entities.Product{
		Name:        req.Name,
		Description: req.Description,
		SKU:         req.SKU,
		Price:       req.Price,
		Currency:    req.Currency,
		CategoryID:  req.CategoryID,
	}
}

func UpdateProductRequestToEntity(req *dto.UpdateProductRequest, product *entities.Product) {
	if req.Name != nil {
		product.Name = *req.Name
	}
	if req.Description != nil {
		product.Description = *req.Description
	}
	if req.SKU != nil {
		product.SKU = *req.SKU
	}
	if req.Price != nil {
		product.Price = *req.Price
	}
	if req.Currency != nil {
		product.Currency = *req.Currency
	}
	if req.CategoryID != nil {
		product.CategoryID = req.CategoryID
	}
	if req.Status != nil {
		product.Status = *req.Status
	}
}
